### PR TITLE
dr/poro

### PR DESCRIPTION
**Pull request by sebwolf-de**
This PR enables dynamic rupture in poroelastic materials. 
Note that the pressure in normal direction is coupled to the fluid pressure when computing the Godunov state. That's why we need the new kernel. For (visco)elastic nothing changes.

### CODE REVIEW

File: postprocessing/validation/compare-energies.py
Comment (davschneller) [inline postprocessing/validation/compare-energies.py:44]: Would be good to see where the problems lie (and how big they are)

File: src/Initializer/InitProcedure/InitSideConditions.cpp
Comment (davschneller) [inline src/Initializer/InitProcedure/InitSideConditions.cpp:99]: It would be good to let it explicitly fail if we don't use poroelastic material

**Review by davschneller — COMMENTED** [review]
Hi, just some comments for in between. I cannot comment that much anyways on the exact nature of the equations right now.

Can we add some files to the formatting or tidying workflow? (e.g. `Equations/poroelastic`?)

File: postprocessing/validation/compare-energies.py
Comment (sebwolf-de) [reply postprocessing/validation/compare-energies.py:44]: If both energies are zero and we compute the relative difference it is (0 - 0) / 0 = NaN.
`NaN < eps` evaluates to false, while `! (NaN > eps)` evaluates to true ;-)

File: src/tests/Reader/NRFReader.t.h
Comment (sebwolf-de) [reply src/tests/Reader/NRFReader.t.h:34]: Sure, but I wanted to wait until after a first review...

File: src/DynamicRupture/FrictionLaws/FrictionSolverCommon.h
Comment (sebwolf-de) [reply src/DynamicRupture/FrictionLaws/FrictionSolverCommon.h:52]: Indeed, I've moved it to CellLocalMatrices.cpp, but forgot to remove it here.

**Comment by sebwolf-de** [comment]
> Can we add some files to the formatting or tidying workflow? (e.g. Equations/poroelastic?)

I'd do that separately, after merging this PR.

File: gitlab-cpu-pipeline.yml
Comment (davschneller) [inline gitlab-cpu-pipeline.yml:47]: Effective duplicate info to line 50 (`pwd`) below?

File: src/DynamicRupture/Typedefs.hpp
Comment (davschneller) [inline src/DynamicRupture/Typedefs.hpp:54]: Maybe `#ifdef USE_POROELASTIC`? Though I'm not sure if it's really necessary here...

File: src/Model/common.hpp
Comment (davschneller) [inline src/Model/common.hpp:211]: Is it good to hard-code that here?

File: src/Model/common.hpp
Comment (davschneller) [inline src/Model/common.hpp:212]: Same here

**Review by davschneller — COMMENTED** [review]
Hi, some more comments. I don't think I can come up with many more, i.e. from my side we'd be good then. Of course, reducing the number of `USE_STP` ifdefs would be good, but that can be something for the future. Maybe we'll have to subclass by solver in the end...

File: src/Equations/poroelastic/Model/PoroelasticSetup.h
Comment (sebwolf-de) [reply src/Equations/poroelastic/Model/PoroelasticSetup.h:257]: The matrices are too bad in the poroelastic case, eigen3 can't handle them. The numerical flux term will become unstable. Somehow LAPACK works way better. That's why poroelastic materials _only_ work with LAPACK and don't work with eigen3.
When I first implemented the poroelastic material model, we decided to have LAPACK as a required dependency only for poroelastic materials and use eigen3 only, whenever possible.

File: src/Initializer/CellLocalMatrices.cpp
Comment (sebwolf-de) [reply src/Initializer/CellLocalMatrices.cpp:588]: Yes, this is correct. In the (visco)elastic case, the `impedanceMatrix` is diagonal, with entries zp, zs.
I'll add a comment to make this more clear. The values are default initialized to zero.

File: src/Initializer/CellLocalMatrices.cpp
Comment (sebwolf-de) [reply src/Initializer/CellLocalMatrices.cpp:164]: Indeed, I guess the call in line 164 ff is obsolete. But that's year old code ;-)

File: src/Initializer/DynamicRupture.h
Comment (sebwolf-de) [reply src/Initializer/DynamicRupture.h:134]: No, we don't always need them, but the overall agreement is to reduce ifdefs wherever possible.
If I add an ifdef here, we would need more ifdefs in CellLocalMatrices.cpp
Otherwise I could create a more complicated class structure with a poroelastic DR LTS tree and a non-poroelastic DR LTS tree, which I would like to avoid. If we add the matrices here and just leave them in the tree, they don't do any harm, IMHO.

File: src/DynamicRupture/FrictionLaws/FrictionSolverCommon.h
Comment (sebwolf-de) [reply src/DynamicRupture/FrictionLaws/FrictionSolverCommon.h:144]: We could, but I don't see a benefit. The impedance matrix is diagonal in the elastic case. Also this would break the GPU implementation of this method.

File: src/DynamicRupture/FrictionLaws/FrictionSolverCommon.h
Comment (sebwolf-de) [reply src/DynamicRupture/FrictionLaws/FrictionSolverCommon.h:310]: Same answer as above ;-)

File: src/DynamicRupture/Typedefs.hpp
Comment (sebwolf-de) [reply src/DynamicRupture/Typedefs.hpp:54]: Same thing with ifdefs, would like to reduce their usage as much as possible. One ifdef in one place, needs five more ifdefs in other places, to make compilation always possible.

File: src/Model/common.hpp
Comment (sebwolf-de) [reply src/Model/common.hpp:211]: IMHO yes. After all it's only executed in debug mode and might help, if a scenario fails.

**Review by davschneller — APPROVED** [review]

**Pull request merged**
